GitPullRequest.Services/AzureDevOpsRepository.cs

```diff
@@ -0,0 +1,66 @@
+using System;
+using System.Linq;
+using System.Text.RegularExpressions;
+using LibGit2Sharp;
+using LibGit2Sharp.Handlers;
+
+namespace GitPullRequest.Services
+{
+    internal class AzureDevOpsRepository : GitRepository
+    {
+
+        public AzureDevOpsRepository(IRepository repo, string remoteName)
+            : base(repo, remoteName)
+        {
+        }
+
+        public override int FindPullRequestForCanonicalName(string canonicalName)
+        {
+            var match = Regex.Match(canonicalName, "^refs/pull/([0-9]+)/merge$");
+            if (match.Success && int.TryParse(match.Groups[1].Value, out int number))
+            {
+                return number;
+            }
+
+            return -1;
+        }
+
+        protected override string GetTipForReference(IRepository repo, Reference reference, CredentialsHandler credentialsHandler)
+        {
+            // Sadly for Azure DevOps there is an automatic merge commit that is at the tip of the ref, so we have to fetch it, in order to find the previous commit
+            // as that is what the users branch will be up to
+            if (reference.CanonicalName.StartsWith("refs/pull/", StringComparison.OrdinalIgnoreCase))
+            {
+                Commit commit;
+                try
```

Comment:
I think you could avoid the try/catch by doing something like this:
```
// Check to see if reference commit is already local
if (repo.Lookup<Commit>(reference.TargetIdentifier) == null)
{
    repo.Network.Fetch(...);
}
```

Reply:
Thanks, that looks much nicer. I was guess, never used LibGit2Sharp before :)

Reply:
You'll find LibGit2Sharp has some slightly strange APIs that return different things based on a generic parameter.